pythran/optimizations/inlining.py

```diff
@@ -0,0 +1,111 @@
+""" Inlining inline functions body. """
+
+from pythran.analyses import Inlinable, Aliases
+from pythran.passmanager import Transformation
+
+import ast
+import copy
+
+
+class Inlining(Transformation):
+
+    """
+    Inline one line functions.
+
+    >>> import ast
+    >>> from pythran import passmanager, backend
+    >>> pm = passmanager.PassManager("test")
+    >>> node = ast.parse('''
+    ... def foo(a, b):
+    ...     return b + b * a
+    ... def bar(b):
+    ...     return foo(2 * b, b) * foo(b, b)''')
+    >>> _, node = pm.apply(Inlining, node)
+    >>> print pm.dump(backend.Python, node)
+    def foo(a, b):
+        return (b + (b * a))
+    def bar(b):
+        _pythran_inline_fooa0 = (2 * b)
+        _pythran_inline_foob0 = b
+        _pythran_inline_fooa1 = b
+        _pythran_inline_foob1 = b
+        return ((_pythran_inline_foob0 + (_pythran_inline_foob0 * \
+_pythran_inline_fooa0)) * (_pythran_inline_foob1 + \
+(_pythran_inline_foob1 * _pythran_inline_fooa1)))
+    """
+
+    def __init__(self):
+        """ fun : Function {name :body} for inlinable functions. """
+        self.update = False
+        self.defs = list()
+        self.call_count = 0
+        super(Inlining, self).__init__(Inlinable, Aliases)
+
+    def visit_Stmt(self, node):
+        """ Add new variable definision before the Statement. """
+        save_defs, self.defs = self.defs or list(), list()
+        self.generic_visit(node)
+        new_defs, self.defs = self.defs, save_defs
+        return new_defs + [node]
+
+    visit_Return = visit_Stmt
+    visit_Assign = visit_Stmt
+    visit_AugAssign = visit_Stmt
+    visit_Print = visit_Stmt
+    visit_For = visit_Stmt
+    visit_While = visit_Stmt
+    visit_If = visit_Stmt
+    visit_With = visit_Stmt
+    visit_Assert = visit_Stmt
+    visit_Expr = visit_Stmt
+
+    def visit_Call(self, node):
+        """
+        Replace function call by inlined function's body.
+
+        We can inline if it alias on only one function.
+        """
+        func_aliases = self.aliases[node.func].aliases
+        if len(func_aliases) == 1:
+            function_def = iter(func_aliases).next()
+            if (isinstance(function_def, ast.FunctionDef) and
+                    function_def.name in self.inlinable):
```

Comment:
My code say : If you have only one alias and it is inlinable , then... and you say : OK but with one alias, you can get not inlinable. It is true, that's why I check both. If you have multiple alias and all of them are inlinable, you can't inline. If you have one alias but not inlinable, you can't inline.